From 90b1adc1f130148a11e1ae11a154edc67bc521f6 Mon Sep 17 00:00:00 2001
From: q66 <daniel@octaforge.org>
Date: Sun, 9 Feb 2020 16:47:57 +0100
Subject: [PATCH] Fix segfault on some platforms

This can be reproduced on e.g. firefox with ppc64 after updating
sqlite from 3.30.1. This is a cherry-pick of an upstream revision
that fixes the problem (https://sqlite.org/src/info/04885763c4cd00cb).
---
 sqlite3.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git sqlite3.c sqlite3.c
index 55dc686..0fb0b55 100644
--- third_party/sqlite3/src/sqlite3.c
+++ third_party/sqlite3/src/sqlite3.c
@@ -121302,12 +121302,14 @@ SQLITE_PRIVATE void sqlite3GenerateConstraintChecks(
             x = *sqlite3VdbeGetOp(v, addrConflictCk);
             if( x.opcode!=OP_IdxRowid ){
               int p2;      /* New P2 value for copied conflict check opcode */
+              const char *zP4;
               if( sqlite3OpcodeProperty[x.opcode]&OPFLG_JUMP ){
                 p2 = lblRecheckOk;
               }else{
                 p2 = x.p2;
               }
-              sqlite3VdbeAddOp4(v, x.opcode, x.p1, p2, x.p3, x.p4.z, x.p4type);
+              zP4 = x.p4type==P4_INT32 ? SQLITE_INT_TO_PTR(x.p4.i) : x.p4.z;
+              sqlite3VdbeAddOp4(v, x.opcode, x.p1, p2, x.p3, zP4, x.p4type);
               sqlite3VdbeChangeP5(v, x.p5);
               VdbeCoverageIf(v, p2!=x.p2);
             }
-- 
2.24.0

